File: rust/crates/openmelon-tui/src/runtime.rs
use anyhow::{Context, Result};
use serde_json::Value;
use std::sync::mpsc;

use crate::llm::{ChatRequest, FinishReason, Message, OpenAIClient, Role, ToolCall, Usage};
use crate::render::{
    flush_markdown_buffer, render_finish_result, render_tool_call, render_tool_result,
    TranscriptMode,
};
use crate::session::Session;
use crate::tools::{ToolEnv, ToolRegistry};

pub struct Runtime {
    pub llm: OpenAIClient,
    pub registry: ToolRegistry,
    pub env: ToolEnv,
    pub max_steps: usize,
    pub reasoning_effort: String,
    pub drain_user_input: Option<Box<dyn FnMut() -> Vec<String> + Send>>,
    pub events: Option<mpsc::Sender<RuntimeEvent>>,
}

#[derive(Debug, Clone)]
pub enum RuntimeEvent {
    TurnStart {
        step: usize,
    },
    TextDelta(String),
    ToolCall(ToolCall),
    ToolResult {
        tool_name: String,
        content: String,
        error: Option<String>,
    },
    TurnEnd {
        step: usize,
        finish: FinishReason,
        usage: Usage,
    },
    QueuedInputApplied {
        count: usize,
    },
}

pub struct RunInput {
    pub system_prompt: String,
    pub user_input: String,
    pub history: Vec<Message>,
}

pub struct RunResult {
    pub messages: Vec<Message>,
    pub steps: usize,
    pub finished: bool,
    pub finish_summary: String,
    pub finish_artifacts: Vec<String>,
}

impl Runtime {
    pub fn run(&mut self, input: RunInput, session: &mut Session) -> Result<RunResult> {
        let mut messages = if input.history.is_empty() {
            let mut seeded = Vec::new();
            if !input.system_prompt.trim().is_empty() {
                seeded.push(Message {
                    role: Role::System,
                    content: input.system_prompt,
                    tool_calls: Vec::new(),
                    tool_call_id: String::new(),
                });
            }
            seeded
        } else {
            input.history
        };
        if !input.user_input.trim().is_empty() {
            messages.push(Message {
                role: Role::User,
                content: input.user_input,
                tool_calls: Vec::new(),
                tool_call_id: String::new(),
            });
        }

        let mut result = RunResult {
            messages: Vec::new(),
            steps: 0,
            finished: false,
            finish_summary: String::new(),
            finish_artifacts: Vec::new(),
        };

        let max_steps = self.max_steps.max(1);
        for step in 0..max_steps {
            result.steps = step + 1;
            let drained = self.drain_user_input();
            if !drained.is_empty() {
                let count = drained.len();
                for text in drained {
                    messages.push(Message {
                        role: Role::User,
                        content: text,
                        tool_calls: Vec::new(),
                        tool_call_id: String::new(),
                    });
                }
                self.emit(RuntimeEvent::QueuedInputApplied { count });
            }
            self.emit(RuntimeEvent::TurnStart { step: step + 1 });
            session.append_event(
                "turn_start",
                serde_json::json!({
                    "step": step + 1,
                    "status": "started",
                }),
            )?;

            let mut markdown_buffer = String::new();
            let events_tx = self.events.clone();
            let response = self
                .llm
                .stream_chat(
                    ChatRequest {
                        messages: messages.clone(),
                        tools: self.registry.specs(),
                        reasoning_effort: self.reasoning_effort.clone(),
                    },
                    |delta| {
                        if let Some(tx) = &events_tx {
                            let _ = tx.send(RuntimeEvent::TextDelta(delta.to_string()));
                        } else {
                            markdown_buffer.push_str(delta);
                            flush_runtime_markdown(&mut markdown_buffer, false);
                        }
                    },
                )
                .with_context(|| format!("runtime chat step {}", step + 1))?;

            if self.events.is_none() {
                flush_runtime_markdown(&mut markdown_buffer, true);
            }
            session.append_event(
                "turn_response",
                serde_json::json!({
                    "step": step + 1,
                    "status": "received",
                    "detail": {
                        "finish": format!("{:?}", response.finish_reason),
                        "tool_calls": response.message.tool_calls.len(),
                        "usage": {
                            "prompt_tokens": response.usage.prompt_tokens,
                            "completion_tokens": response.usage.completion_tokens,
                            "total_tokens": response.usage.total_tokens,
                        }
                    }
                }),
            )?;

            let tool_calls = response.message.tool_calls.clone();
            let finish_reason = response.finish_reason;
            let usage = response.usage;
            messages.push(response.message);
            if tool_calls.is_empty() {
                self.emit(RuntimeEvent::TurnEnd {
                    step: step + 1,
                    finish: finish_reason,
                    usage,
                });
                result.finished = matches!(finish_reason, FinishReason::Stop | FinishReason::Other);
                result.messages = messages;
                return Ok(result);
            }

            for call in tool_calls {
                session.append_event(
                    "tool_call",
                    serde_json::json!({
                        "step": step + 1,
                        "tool": call.name,
                        "status": "started",
                        "detail": { "arguments": call.arguments },
                    }),
                )?;
                if call.name != "finish" {
                    if self.events.is_some() {
                        self.emit(RuntimeEvent::ToolCall(call.clone()));
                    } else {
                        flush_runtime_markdown(&mut markdown_buffer, true);
                        println!();
                        println!("{}", render_tool_call(&call, TranscriptMode::Styled));
                    }
                }
                let (content, dispatch_err) =
                    match self
                        .registry
                        .dispatch(&self.env, &call.name, call.arguments.clone())
                    {
                        Ok(value) => (value, None),
                        Err(err) => (
                            serde_json::json!({ "error": err.to_string() }),
                            Some(err.to_string()),
                        ),
                    };
                if let Some(err) = dispatch_err {
                    let content_for_render =
                        serde_json::json!({ "error": err.clone() }).to_string();
                    if self.events.is_some() {
                        self.emit(RuntimeEvent::ToolResult {
                            tool_name: call.name.clone(),
                            content: content_for_render.clone(),
                            error: Some(err.clone()),
                        });
                    } else {
                        println!(
                            "{}",
                            render_tool_result(
                                &call.name,
                                &content_for_render,
                                TranscriptMode::Styled,
                            )
                        );
                    }
                    session.append_event(
                        "tool_result",
                        serde_json::json!({
                            "step": step + 1,
                            "tool": call.name,
                            "status": "error",
                            "detail": { "error": err },
                        }),
                    )?;
                } else if call.name == "finish" {
                    if let Some(summary) = content.get("summary").and_then(Value::as_str) {
                        result.finish_summary = summary.to_string();
                    }
                    if let Some(artifacts) = content.get("artifacts").and_then(Value::as_array) {
                        result.finish_artifacts = artifacts
                            .iter()
                            .filter_map(|v| v.as_str().map(ToString::to_string))
                            .collect();
                    }
                    let finish = serde_json::to_string(&content)?;
                    if self.events.is_some() {
                        self.emit(RuntimeEvent::ToolResult {
                            tool_name: call.name.clone(),
                            content: finish.clone(),
                            error: None,
                        });
                    } else {
                        let rendered = render_finish_result(&finish, 88, TranscriptMode::Styled);
                        if !rendered.trim().is_empty() {
                            println!("{rendered}");
                        }
                    }
                    session.append_event(
                        "tool_result",
                        serde_json::json!({
                            "step": step + 1,
                            "tool": call.name,
                            "status": "ok",
                            "detail": content,
                        }),
                    )?;
                } else {
                    let content_for_render = serde_json::to_string(&content)?;
                    if self.events.is_some() {
                        self.emit(RuntimeEvent::ToolResult {
                            tool_name: call.name.clone(),
                            content: content_for_render.clone(),
                            error: None,
                        });
                    } else {
                        println!(
                            "{}",
                            render_tool_result(
                                &call.name,
                                &content_for_render,
                                TranscriptMode::Styled,
                            )
                        );
                    }
                    session.append_event(
                        "tool_result",
                        serde_json::json!({
                            "step": step + 1,
                            "tool": call.name,
                            "status": if content.get("error").is_some() { "error" } else { "ok" },
                            "detail": content,
                        }),
                    )?;
                }

                messages.push(Message {
                    role: Role::Tool,
                    content: serde_json::to_string(&content)?,
                    tool_calls: Vec::new(),
                    tool_call_id: call.id,
                });

                if call.name == "finish" {
                    self.emit(RuntimeEvent::TurnEnd {
                        step: step + 1,
                        finish: finish_reason,
                        usage,
                    });
                    result.finished = true;
                    result.messages = messages;
                    return Ok(result);
                }
            }
            self.emit(RuntimeEvent::TurnEnd {
                step: step + 1,
                finish: finish_reason,
                usage,
            });
        }

        result.messages = messages;
        Ok(result)
    }

    fn drain_user_input(&mut self) -> Vec<String> {
        let Some(drain) = &mut self.drain_user_input else {
            return Vec::new();
        };
        drain()
            .into_iter()
            .map(|text| text.trim().to_string())
            .filter(|text| !text.is_empty())
            .collect()
    }

    fn emit(&self, event: RuntimeEvent) {
        if let Some(tx) = &self.events {
            let _ = tx.send(event);
        }
    }
}

fn flush_runtime_markdown(buffer: &mut String, force: bool) {
    let Some(rendered) = flush_markdown_buffer(buffer, force, 88, TranscriptMode::Styled) else {
        return;
    };
    if rendered.trim().is_empty() {
        return;
    }
    println!("{rendered}");
    println!();
    let _ = std::io::Write::flush(&mut std::io::stdout());
}

File: tui/src/engine/runtime.ts
// openmelon's tool-driven agent loop, ported from internal/runtime.
//
// A classic ReAct cycle:
//   1. Send (system prompt, conversation, tools) to the LLM.
//   2. The model replies with text and/or tool_calls.
//   3. Dispatch each tool_call via the registry, append the result as a
//      tool message.
//   4. Loop until: the model finishes naturally, calls the `finish` tool,
//      or hits maxSteps.
//
// Provider-agnostic: anything implementing LLMClient works. When the client
// also implements streamChat, the loop streams text so the TUI renders as it
// arrives.

import type {ChatRequest, FinishReason, LLMClient, Message, Tool, ToolCall, Usage} from './llm/types.js';

export const defaultMaxSteps = 16;

/** A tool's machine-readable contract sent to the model. */
export type ToolSpec = {
	name: string;
	description: string;
	/** JSON Schema object. */
	parameters: Record<string, unknown>;
};

/**
 * The runtime depends only on this narrow registry surface. The tools package
 * implements it (specs to advertise + dispatch to run).
 */
export type ToolRegistry = {
	specs(): ToolSpec[];
	/** Run the named tool with the raw JSON argument string; resolve a JSON-serializable value or throw. */
	dispatch(name: string, argsJson: string, signal?: AbortSignal): Promise<unknown>;
};

/** Structured per-turn events. Implementations render however they like. */
export type Tracer = {
	onTurnStart?(turn: number): void;
	/** Streamed when the client supports it; otherwise fired once with the full text. */
	onText?(delta: string): void;
	onToolCall?(call: ToolCall): void;
	onToolResult?(call: ToolCall, content: string, err: Error | null): void;
	onTurnEnd?(turn: number, finish: FinishReason, usage: Usage): void;
};

export type RunInput = {
	/** Sets the agent's behavior + project context. Sent only when history is empty. */
	systemPrompt?: string;
	/** The user's request for this run. Appended after history. */
	userInput?: string;
	/** Prior conversation (incl. tool messages). When non-empty, systemPrompt is ignored (system already at [0]). */
	history?: Message[];
	temperature?: number;
	maxTokens?: number;
};

export type RunResult = {
	/** Full conversation incl. all tool calls + replies. Pass back as history to continue. */
	messages: Message[];
	/** LLM round-trips taken in this run (not cumulative). */
	steps: number;
	/** True when the loop exited via `finish` or a natural stop. */
	finished: boolean;
	finishSummary?: string;
	finishArtifacts?: string[];
};

export type RuntimeOptions = {
	llm: LLMClient;
	registry: ToolRegistry;
	tracer?: Tracer;
	maxSteps?: number;
	reasoningEffort?: string;
	/** Called before each model request to fold in user corrections queued mid-run. */
	drainUserInput?: () => string[];
};

export class Runtime {
	constructor(private readonly opts: RuntimeOptions) {}

	get reasoningEffort(): string {
		return this.opts.reasoningEffort ?? '';
	}

	async run(input: RunInput, signal?: AbortSignal): Promise<RunResult> {
		const {llm, registry, tracer} = this.opts;
		const maxSteps = this.opts.maxSteps && this.opts.maxSteps > 0 ? this.opts.maxSteps : defaultMaxSteps;

		const wireTools: Tool[] = registry.specs().map(s => ({
			name: s.name,
			description: s.description,
			parameters: s.parameters
		}));

		// Seed the message list: new conversations get system + user;
		// continuations get history + user.
		let messages: Message[] = [];
		if (input.history && input.history.length > 0) {
			messages = [...input.history];
		} else if (input.systemPrompt) {
			messages.push({role: 'system', content: input.systemPrompt});
		}
		if (input.userInput) {
			messages.push({role: 'user', content: input.userInput});
		}

		const out: RunResult = {messages, steps: 0, finished: false};

		for (let step = 0; step < maxSteps; step++) {
			out.steps = step + 1;
			messages = appendDrained(messages, this.opts.drainUserInput);
			tracer?.onTurnStart?.(step + 1);

			const req: ChatRequest = {
				messages,
				tools: wireTools,
				temperature: input.temperature,
				maxTokens: input.maxTokens,
				reasoningEffort: this.reasoningEffort
			};

			let resp;
			if (llm.streamChat) {
				resp = await llm.streamChat(req, {onText: d => tracer?.onText?.(d)}, signal);
			} else {
				resp = await llm.chat(req, signal);
				if (resp.message.content) {
					tracer?.onText?.(resp.message.content);
				}
			}

			messages = [...messages, resp.message];
			const toolCalls = resp.message.toolCalls ?? [];

			if (toolCalls.length === 0) {
				tracer?.onTurnEnd?.(step + 1, resp.finishReason, resp.usage);
				out.messages = messages;
				out.finished = resp.finishReason === 'stop' || resp.finishReason === 'other';
				return out;
			}

			let hitFinish = false;
			for (const tc of toolCalls) {
				tracer?.onToolCall?.(tc);
				let content: string;
				let dispatchErr: Error | null = null;
				let resValue: unknown = null;
				try {
					resValue = await registry.dispatch(tc.name, tc.arguments, signal);
					content = safeStringify(resValue);
					dispatchErr = toolContentError(content);
				} catch (error) {
					dispatchErr = error as Error;
					content = JSON.stringify({error: (error as Error).message});
				}
				tracer?.onToolResult?.(tc, content, dispatchErr);
				messages = [...messages, {role: 'tool', toolCallId: tc.id, content}];

				if (tc.name === 'finish' && !dispatchErr) {
					const m = resValue as {summary?: unknown; artifacts?: unknown} | null;
					if (m && typeof m === 'object') {
						if (typeof m.summary === 'string' && m.summary) {
							out.finishSummary = m.summary;
						}
						if (Array.isArray(m.artifacts)) {
							out.finishArtifacts = m.artifacts.filter((a): a is string => typeof a === 'string');
						}
					}
					hitFinish = true;
				}
			}

			tracer?.onTurnEnd?.(step + 1, resp.finishReason, resp.usage);
			if (hitFinish) {
				out.messages = messages;
				out.finished = true;
				return out;
			}
		}

		out.messages = messages;
		throw new RuntimeMaxStepsError(maxSteps, out);
	}
}

/** Thrown when the loop hits maxSteps without finishing. Carries the partial result. */
export class RuntimeMaxStepsError extends Error {
	constructor(
		readonly maxSteps: number,
		readonly partial: RunResult
	) {
		super(`runtime: hit maxSteps=${maxSteps} without finishing`);
		this.name = 'RuntimeMaxStepsError';
	}
}

function appendDrained(messages: Message[], drain: (() => string[]) | undefined): Message[] {
	if (!drain) {
		return messages;
	}
	const extra = drain()
		.map(t => t.trim())
		.filter(Boolean)
		.map((content): Message => ({role: 'user', content}));
	return extra.length > 0 ? [...messages, ...extra] : messages;
}

function safeStringify(value: unknown): string {
	try {
		const s = JSON.stringify(value ?? null);
		return s ?? 'null';
	} catch (error) {
		return JSON.stringify({error: `tool result not serializable: ${(error as Error).message}`});
	}
}

/**
 * If a tool returned a JSON object with a non-empty `error` field, surface it
 * as an Error so the tracer marks it failed — while the content is still fed
 * back to the model so it can self-correct.
 */
function toolContentError(content: string): Error | null {
	let obj: unknown;
	try {
		obj = JSON.parse(content);
	} catch {
		return null;
	}
	if (!obj || typeof obj !== 'object' || !('error' in obj)) {
		return null;
	}
	const raw = (obj as {error: unknown}).error;
	if (raw == null) {
		return null;
	}
	if (typeof raw === 'string') {
		return raw.trim() ? new Error(raw) : null;
	}
	return new Error(JSON.stringify(raw));
}
